parser/src/json/schema.rs

```diff
@@ -192,11 +192,252 @@ impl Schema {
+                    },
+                    required,
+                }
+            }
+            //TODO: get types for error message
+            _ => Schema::Unsatisfiable {
+                reason: "incompatible types".to_string(),
+            },
+        };
+        Ok(merged.normalize())
+    }
+
+    fn is_verifiably_disjoint_from(&self, other: &Schema) -> bool {
+        match (self, other) {
+            (Schema::Unsatisfiable { .. }, _) => true,
+            (_, Schema::Unsatisfiable { .. }) => true,
+            (Schema::Any, _) => false,
+            (_, Schema::Any) => false,
+            (Schema::Boolean, Schema::LiteralBool { .. }) => false,
+            (Schema::LiteralBool { .. }, Schema::Boolean) => false,
+            (Schema::Ref { .. }, _) => false, // TODO: could resolve
+            (_, Schema::Ref { .. }) => false, // TODO: could resolve
+            (Schema::LiteralBool { value: value1 }, Schema::LiteralBool { value: value2 }) => {
+                value1 != value2
+            }
+            (Schema::AnyOf { options }, _) => options
+                .iter()
+                .all(|opt| opt.is_verifiably_disjoint_from(other)),
+            (_, Schema::AnyOf { options }) => options
+                .iter()
+                .all(|opt| self.is_verifiably_disjoint_from(opt)),
+            (Schema::OneOf { options }, _) => options
+                .iter()
+                .all(|opt| opt.is_verifiably_disjoint_from(other)),
+            (_, Schema::OneOf { options }) => options
+                .iter()
+                .all(|opt| self.is_verifiably_disjoint_from(opt)),
+            // TODO: could actually compile the regexes and check for overlap
+            (
+                Schema::String {
+                    regex: Some(RegexAst::Literal(lit1)),
+                    ..
+                },
+                Schema::String {
+                    regex: Some(RegexAst::Literal(lit2)),
+                    ..
+                },
+            ) => lit1 != lit2,
+            (
+                Schema::Object {
+                    properties: props1,
+                    required: req1,
+                    additional_properties: add1,
+                },
+                Schema::Object {
+                    properties: props2,
+                    required: req2,
+                    additional_properties: add2,
+                },
+            ) => req1.union(req2).any(|key| {
+                let prop1 = props1
+                    .get(key)
+                    .unwrap_or(add1.as_deref().unwrap_or(&Schema::Any));
+                let prop2 = props2
+                    .get(key)
+                    .unwrap_or(add2.as_deref().unwrap_or(&Schema::Any));
+                prop1.is_verifiably_disjoint_from(prop2)
+            }),
+            _ => {
+                // Except for in the cases above, it should suffice to check that the types are different
+                mem::discriminant(self) != mem::discriminant(other)
```

Comment:
Ha! I didn't know this existed!

Actually, on that note, maybe in future, Schema::Boolean should have value: Optional<bool> and we could drop LiteralBool? The same way numbers and strings are treated.

Reply:
Oh, duh. That's a good idea. 😂

Reply:
@mmoskal I think we really need to implement refs here (could be in the next PR), but we'll definitely have to be careful about ref cycles causing infinite loops. Maybe setting a max ref depth on this method would be an ok solution

Reply:
P.s. matching on RegexAst::Literal can replace the const_string check in your const_compile PR.

Reply:
I think we don't need to match on const anything - we just always compile strings as regexes and or them together with other options in AnyOf